### PR TITLE
Refactor chainRunType references

### PR DESCRIPTION
**Pull request by axenteoctavian**
## Reasoning behind the pull request
- `chainRunType` refactor
- 
- 
  
## Proposed changes
- Refactor remaining `chainRunType` references with factories
- 
- 

## Testing procedure
- running sovereign node
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by mariusmihaic — CHANGES_REQUESTED** [review]
This is wrong.
1. Do not create a PR which includes another merges/prs
2. Split refactor into multiple, smaller, targeted prs

File: factory/vm/sovereignVmContainerShardCreator.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// NewSovereignVmContainerShardFactory creates a new sovereign vm container shard factory
+func NewSovereignVmContainerShardFactory(bhhc hooks.BlockChainHookHandlerCreator) (*sovereignVmContainerShardFactory, error) {
+	return &sovereignVmContainerShardFactory{
+		blockChainHookHandlerCreator: bhhc,
+	}, nil
+}
+
+// CreateVmContainerFactoryShard will create a new instance of sovereign vm container and factory for shard
+func (svcsf *sovereignVmContainerShardFactory) CreateVmContainerFactoryShard(argsHook hooks.ArgBlockChainHook, args ArgsVmContainerFactory) (process.VirtualMachinesContainer, process.VirtualMachinesContainerFactory, error) {
+	blockChainHookImpl, err := svcsf.blockChainHookHandlerCreator.CreateBlockChainHookHandler(argsHook)
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator.go:25]: Actually, sovereign container should hold internally both shard+meta creators(from constructor, as pointers)
Inside this func it should call:
1. shard.Create...
2. meta.Create...

File: factory/vm/sovereignVmContainerShardCreator.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	sovereignFactory, err := NewSovereignVmContainerMetaFactory(svcsf.blockChainHookHandlerCreator)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	metaStorage := argsHook.ConfigSCStorage
+	metaStorage.DB.FilePath = metaStorage.DB.FilePath + "_meta"
+	argsHook.ConfigSCStorage = metaStorage
+	argsHook.Counter = counters.NewDisabledCounter()
+	argsHook.MissingTrieNodesNotifier = syncer.NewMissingTrieNodesNotifier()
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator.go:61]: You should not create new pointers, but used the same ones as in other components.
Anyway, this should probably be removed after the refactor above is done

File: factory/vm/sovereignVmContainerShardCreator.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+type sovereignVmContainerShardFactory struct {
+	blockChainHookHandlerCreator hooks.BlockChainHookHandlerCreator
+}
+
+// NewSovereignVmContainerShardFactory creates a new sovereign vm container shard factory
+func NewSovereignVmContainerShardFactory(bhhc hooks.BlockChainHookHandlerCreator) (*sovereignVmContainerShardFactory, error) {
+	return &sovereignVmContainerShardFactory{
+		blockChainHookHandlerCreator: bhhc,
+	}, nil
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator.go:20]: Check nil and return error

File: factory/vm/sovereignVmContainerShardCreator_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/multiversx/mx-chain-go/factory/vm"
+	factory2 "github.com/multiversx/mx-chain-go/process/factory"
+	"github.com/multiversx/mx-chain-go/testscommon/factory"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewSovereignVmContainerShardCreatorFactory(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator_test.go:13]: Also add test for nil input pointer

File: factory/vm/vmContainerMetaCreator.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+type vmContainerMetaFactory struct {
+	blockChainHookHandlerCreator hooks.BlockChainHookHandlerCreator
+}
+
+// NewVmContainerMetaFactory creates a new vm container meta factory
+func NewVmContainerMetaFactory(bhhc hooks.BlockChainHookHandlerCreator) (*vmContainerMetaFactory, error) {
+	return &vmContainerMetaFactory{
+		blockChainHookHandlerCreator: bhhc,
+	}, nil
```
Comment (mariusmihaic) [inline factory/vm/vmContainerMetaCreator.go:17]: Check nil + return error

File: factory/vm/vmContainerMetaCreator_test.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		},
+		ValidatorAccountsDB: &stateMock.AccountsStub{},
+		UserAccountsDB:      &stateMock.AccountsStub{},
+		ChanceComputer:      &mock.RaterMock{},
+		ShardCoordinator:    &mock.ShardCoordinatorStub{},
+		EnableEpochsHandler: enableEpochsHandlerMock.NewEnableEpochsHandlerStub(common.StakeFlag),
+	}
+}
+
+func TestNewVmContainerMetaCreatorFactory(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/vm/vmContainerMetaCreator_test.go:121]: Also test for nil here

File: factory/vm/vmContainerShardCreator.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+type vmContainerShardFactory struct {
+	blockChainHookHandlerCreator hooks.BlockChainHookHandlerCreator
+}
+
+// NewVmContainerShardFactory creates a new vm container shard factory
+func NewVmContainerShardFactory(bhhc hooks.BlockChainHookHandlerCreator) (*vmContainerShardFactory, error) {
+	return &vmContainerShardFactory{
+		blockChainHookHandlerCreator: bhhc,
+	}, nil
```
Comment (mariusmihaic) [inline factory/vm/vmContainerShardCreator.go:17]: Check nil + return error

File: factory/vm/vmContainerShardCreator_test.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		EnableEpochsHandler: &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		WasmVMChangeLocker:  &sync.RWMutex{},
+		ESDTTransferParser:  esdtTransferParser,
+		BuiltInFunctions:    vmcommonBuiltInFunctions.NewBuiltInFunctionContainer(),
+		BlockChainHook:      &testscommon.BlockChainHookStub{},
+		Hasher:              &hashingMocks.HasherMock{},
+	}
+}
+
+func TestNewVmContainerShardCreatorFactory(t *testing.T) {
```
Comment (mariusmihaic) [inline factory/vm/vmContainerShardCreator_test.go:50]: Also test here

File: factory/vm/vmContainerShardCreator_test.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/multiversx/mx-chain-vm-common-go/parsers"
+	wasmConfig "github.com/multiversx/mx-chain-vm-go/config"
+	"github.com/stretchr/testify/require"
+)
+
+func makeVMConfig() config.VirtualMachineConfig {
+	return config.VirtualMachineConfig{
+		WasmVMVersions: []config.WasmVMVersionByEpoch{
+			{StartEpoch: 0, Version: "v1.2"},
+			{StartEpoch: 10, Version: "v1.2"},
```
Comment (mariusmihaic) [inline factory/vm/vmContainerShardCreator_test.go:27]: Do you really need all of these?

**Review by mariusmihaic — CHANGES_REQUESTED** [review]

File: factory/api/apiResolverFactory_test.go [hunk 10]
```diff
@@ -214,63 +238,63 @@ func TestCreateApiResolver(t *testing.T) {
...
 		failingStepsInstance.reset()
-		failingStepsInstance.enableEpochsHandlerFailingStep = 4
+		failingStepsInstance.enableEpochsHandlerFailingStep = 6
 		apiResolver, err := api.CreateApiResolver(failingArgs)
 		require.NotNil(t, err)
 		require.True(t, strings.Contains(strings.ToLower(err.Error()), "enable epochs handler"))
 		require.True(t, check.IfNil(apiResolver))
 	})
-	t.Run("NewTransactionCostEstimator fails should error", func(t *testing.T) {
+	t.Run("NewFeeComputer fails should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (axenteoctavian) [reply factory/api/apiResolverFactory_test.go:247]: First, the name `NewTransactionCostEstimator` might be very old.
In Radu's PR this function was like [this](https://github.com/multiversx/mx-chain-go/blob/7d8df22022526e394a8c8cac828106afef155bf6/factory/api/apiResolverFactory.go#L207) , so `failingStepsInstance.enableEpochsHandlerFailingStep` is correct. But right now the only parameter is `EconomicsData` so I had to change

File: factory/processing/blockProcessorCreator.go [hunk 11]
```diff
@@ -144,26 +145,57 @@ func (pcf *processComponentsFactory) newShardBlockProcessor(
...
-		pcf.state.AccountsAdapter(),
-		missingTrieNodesNotifier,
-		builtInFuncFactory.BuiltInFunctionContainer(),
-		esdtTransferParser,
-		wasmVMChangeLocker,
-		pcf.config.SmartContractsStorage,
-		builtInFuncFactory.NFTStorageHandler(),
-		builtInFuncFactory.ESDTGlobalSettingsHandler(),
-	)
+	counter, err := counters.NewUsageCounter(esdtTransferParser)
```
[the remark this answers is not in the document]
Comment (axenteoctavian) [reply factory/processing/blockProcessorCreator.go:148]: This counter is for `VmContainerCreator factory`.
I refactored the `createVMFactoryShard` and `createVMFactoryMeta` and used the factory instead (this was for other file)

File: factory/api/apiResolverFactory.go [hunk 12]
```diff
@@ -3,6 +3,7 @@ package api
 import (
 	"errors"
 	"fmt"
+	factoryVm "github.com/multiversx/mx-chain-go/factory/vm"
```
Comment (sasurobert) [inline factory/api/apiResolverFactory.go:6]: goimports

File: factory/vm/vmContainerMetaCreator_test.go [hunk 13]
```diff
@@ -0,0 +1,174 @@
+package vm_test
+
+import (
+	"fmt"
+	"github.com/multiversx/mx-chain-go/process"
```
Comment (sasurobert) [inline factory/vm/vmContainerMetaCreator_test.go:5]: goimports

File: factory/vm/vmContainerShardCreator_test.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
+package vm_test
+
+import (
+	"fmt"
+	"github.com/multiversx/mx-chain-go/process"
```
Comment (sasurobert) [inline factory/vm/vmContainerShardCreator_test.go:5]: goimports

File: testscommon/factory/vmContainerMetaFactoryMock.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package factory
+
+import (
+	factoryVm "github.com/multiversx/mx-chain-go/factory/vm"
+	"github.com/multiversx/mx-chain-go/integrationTests/mock"
+	"github.com/multiversx/mx-chain-go/process"
+	"github.com/multiversx/mx-chain-go/process/smartContract/hooks"
+)
+
+type VMContainerMetaFactoryMock struct {
```
Comment (sasurobert) [inline testscommon/factory/vmContainerMetaFactoryMock.go:10]: comments on public functions

File: testscommon/factory/vmContainerShardFactoryMock.go [hunk 16]
```diff
@@ -0,0 +1,23 @@
+package factory
+
+import (
+	factoryVm "github.com/multiversx/mx-chain-go/factory/vm"
+	"github.com/multiversx/mx-chain-go/integrationTests/mock"
+	"github.com/multiversx/mx-chain-go/process"
+	"github.com/multiversx/mx-chain-go/process/smartContract/hooks"
+)
+
+type VMContainerShardFactoryMock struct {
```
Comment (sasurobert) [inline testscommon/factory/vmContainerShardFactoryMock.go:10]: comments on public functions

File: factory/vm/sovereignVmContainerShardCreator.go [hunk 17]
```diff
@@ -3,64 +3,48 @@ package vm
...
-		blockChainHookHandlerCreator: bhhc,
-	}, nil
-}
-
-// CreateVmContainerFactoryShard will create a new instance of sovereign vm container and factory for shard
-func (svcsf *sovereignVmContainerShardFactory) CreateVmContainerFactoryShard(argsHook hooks.ArgBlockChainHook, args ArgsVmContainerFactory) (process.VirtualMachinesContainer, process.VirtualMachinesContainerFactory, error) {
-	blockChainHookImpl, err := svcsf.blockChainHookHandlerCreator.CreateBlockChainHookHandler(argsHook)
-	if err != nil {
-		return nil, nil, err
+func NewSovereignVmContainerShardFactory(bhhc hooks.BlockChainHookHandlerCreator, vcm VmContainerCreator, vcs VmContainerCreator) (*sovereignVmContainerShardFactory, error) {
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator.go:16]: Can we please use full name/detailed naming for vcm/vcs variables?

File: factory/vm/sovereignVmContainerShardCreator_test.go [hunk 18]
```diff
@@ -13,17 +15,50 @@ import (
...
+		runTypeComponents := componentsMock.GetRunTypeComponents()
+		sovereignVmContainerShardFactory, err := vm.NewSovereignVmContainerShardFactory(bhhc, runTypeComponents.VmContainerMetaFactoryCreator(), runTypeComponents.VmContainerShardFactoryCreator())
+		require.Nil(t, err)
+		require.False(t, sovereignVmContainerShardFactory.IsInterfaceNil())
+	})
+
+	t.Run("nil blockChainHookHandlerCreator", func(t *testing.T) {
+		t.Parallel()
+
+		sovereignVmContainerShardFactory, err := vm.NewSovereignVmContainerShardFactory(nil, nil, nil)
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator_test.go:30]: Please use nil only for first argument(blockchainHool) and mocks/components for the rest

File: factory/vm/sovereignVmContainerShardCreator.go [hunk 19]
```diff
@@ -3,64 +3,48 @@ package vm
...
+	if bhhc == nil {
```
Comment (mariusmihaic) [inline factory/vm/sovereignVmContainerShardCreator.go:17]: Please use check.IfNil for all checks.
Applies for all files

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]
